### PR TITLE
add multi-stage builds to minimize docker image size.

### PR DESCRIPTION
**Pull request by voldemorte**
#78 

The resultant image size is that of about 80MB.

Signed-off-by: Nikhil Chaudhari nikhil.c@mailbox.org

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -14,5 +14,8 @@ WORKDIR /pwru
 COPY . .
 RUN make
 RUN chmod a+x /pwru
-ENTRYPOINT ["./pwru"]
-CMD ["./pwru", "-output-tuple"]
+
+FROM ubuntu:20.04 AS base
+COPY --from=build /pwru/pwru /usr/local/bin/
+ENTRYPOINT ["/usr/local/bin/pwru"]
+CMD ["/usr/local/bin/pwru", "-output-tuple"]
```
Comment (brb) [inline Dockerfile:21]: Let's drop `CMD`. AFAIK, this then would just run `pwru`.

**Review by brb — CHANGES_REQUESTED** [review]
Thanks! Small nit.

File: Dockerfile [hunk 2]
```diff
@@ -1,4 +1,4 @@
-FROM ubuntu:20.04
+FROM ubuntu:20.04 AS build
```
Comment (brb) [inline Dockerfile:1]: While here, we can s/20.04/22.04/.

File: Dockerfile (hunk 1 above)
Comment (voldemorte) [reply Dockerfile]: Yes dropping CMD would just run the ENTRYPOINT. Have removed CMD.

File: Dockerfile [hunk 3]
```diff
@@ -14,5 +14,7 @@ WORKDIR /pwru
 COPY . .
 RUN make
 RUN chmod a+x /pwru
-ENTRYPOINT ["./pwru"]
-CMD ["./pwru", "-output-tuple"]
+
+FROM ubuntu:22.04 AS base
```
Comment (brb) [inline Dockerfile:18]:
```suggestion
FROM scratch
```
This reduces the final image size to 4MB.

File: Dockerfile [hunk 4]
```diff
@@ -1,4 +1,4 @@
-FROM ubuntu:20.04
+FROM ubuntu:22.04 AS build
```
Comment (brb) [inline Dockerfile:1]: Could you add `git` to `RUN apt install -y wget gnupg2`? This will allow us to add the proper version to the final binary.

**Review by brb — CHANGES_REQUESTED** [review]
Two more comments. Also, could you squash all your commits into single one, and add `Signed-off-by` to the final commit msg?

File: Dockerfile [hunk 5]
```diff
@@ -1,4 +1,4 @@
...
 RUN apt update -y -q
```
Comment (tklauser) [inline Dockerfile:2]: While at it, could you merge these `RUN` instructions (i.e. `RUN apt update -y -q && DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -y -q curl build-essential ca-certificates ...`) to reduce the number of layers (see https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers) as well?

File: Dockerfile (hunk 3 above)
Comment (voldemorte) [reply Dockerfile]: Yes this definitely reduced the size of the docker image. This was something new I learnt today, Thanks. :)

File: Dockerfile (hunk 4 above)
Comment (voldemorte) [reply Dockerfile:1]: I've added `git` to the `apt install`.

File: Dockerfile (hunk 5 above)
Comment (voldemorte) [reply Dockerfile]: I've managed to club the different `RUN` steps to reduce the number of layers.

**Review by brb — APPROVED** [review]
Thanks!

**Pull request merged**